### ISSUE TITLE
Compiler crash when unpacking and returning optionals

### ISSUE THREAD

**Issue by Tetralux** [ziglang/zig#3242]
```zig
const std = @import("std");
const warn = std.debug.warn;

const T = struct{};

const U = struct {
    str: []const u8,

    fn f(self: *U) ?T {
        var str = self.str;
        return str[0..1];
    }

    fn g(self: *U) ?T {
        if (self.f()) |e| return e;
        return null;
    }
};

pub fn main() void { // return type can also be 'anyerror!void' as well -- still crashes.
    var u = U{ .str = "" };

    // Either of these crashes the compiler.
    // if (u.f()) |e| return e; 
    // if (u.g()) |e| warn("{}\n", e);
    
    warn("done\n");
}
```

**Comment by Tetralux** [comment]
There's a related problem that causes invalid LLVM IR to be generated, which is what I was originally trying to reproduce when I came across this; this outright crashes rather than generating bad IR though.

**Comment by Vexu** [comment]
Fixed by 5ea79bfc4a0a1269930d98faee36a8f6cb0b0401
```
./build/a.zig:45:27: error: expected type 'void', found 'T'
    if (u.f()) |e| return e; 
                          ^
./build/a.zig:25:11: note: T declared here
const T = struct{};
          ^
./build/a.zig:32:19: error: expected type 'T', found '[]const u8'
        return str[0..1];
                  ^
./build/a.zig:25:11: note: T declared here
const T = struct{};
          ^
```
